pinot-common/src/main/java/org/apache/pinot/common/function/scalar/CaseWhenScalarFunction.java

```diff
@@ -0,0 +1,76 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.common.function.scalar;
+
+import javax.annotation.Nullable;
+import org.apache.pinot.common.function.FunctionInfo;
+import org.apache.pinot.common.function.PinotScalarFunction;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.spi.annotations.ScalarFunction;
+
+
+@ScalarFunction(nullableParameters = true, names = {"case", "caseWhen"})
+public class CaseWhenScalarFunction implements PinotScalarFunction {
+
+  private static final FunctionInfo FUNCTION_INFO;
+
+  static {
+    try {
+      FUNCTION_INFO = new FunctionInfo(CaseWhenScalarFunction.class.getMethod("caseWhen", Object[].class),
+          CaseWhenScalarFunction.class, true);
+    } catch (NoSuchMethodException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public String getName() {
+    return "case";
+  }
+
+  @Nullable
+  @Override
+  public FunctionInfo getFunctionInfo(DataSchema.ColumnDataType[] argumentTypes) {
+    // This function is an Object based function that can handle any type of arguments. Ideally, we would at least
+    // check if the arguments are of type BOOLEAN, OBJECT, BOOLEAN, OBJECT, ... but that would break backward
+    // compatibility with the v1 engine since the earlier implementation did not have such a check. In the v2 engine,
+    // the standard Calcite CASE operator has the necessary type checks.
+    return getFunctionInfo(argumentTypes.length);
+  }
+
+  @Nullable
+  @Override
+  public FunctionInfo getFunctionInfo(int numArguments) {
+    if (numArguments < 2) {
+      return null;
+    }
+
+    return FUNCTION_INFO;
+  }
+
+  public static Object caseWhen(Object... objs) {
```

Comment:
In order to align the behavior with transform function, do we need to calculate the result type and do a cast in the end?

Reply:
In v2, will the `SqlOperator` derived the correct return type, and then cast the return value to the desired type? If so, then we get consistent behavior in v2

Reply:
Yes, the `SqlCaseOperator` has similar logic to determine the "least restrictive" type across all the `then` clauses and the `else` clause and that is determined as the return type for the operator - https://github.com/apache/calcite/blob/78e873d39c0364f9f36055b9cbe0600dfad49c71/core/src/main/java/org/apache/calcite/sql/fun/SqlCaseOperator.java#L219-L323. We then cast the actual returned value from the scalar function to that computed type (or the Pinot equivalent type for the Calcite `RelDataType`) in the v2 engine. So it's only the v1 engine where the scalar function return type won't be determined based on the input operand types.